pallets/account/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+        // ensure account_id and address has not been mapped
+        ensure!(
+            !EvmAccounts::<T>::contains_key(&account_id),
+            Error::<T>::AccountIdHasMapped
+        );
+        // This is not required since checking one mapping is sufficent
+        // but this is just for sanity check
+        if let Some(addr) = address {
+            ensure!(
+                !NativeAccounts::<T>::contains_key(addr),
+                Error::<T>::EthAddressHasMapped
+            );
+        }
+        Ok(())
+    }
+
+    /// Add the given pair to create double mappings
+    fn add_mappings(account_id: T::AccountId, address: EvmAddress) {
+        NativeAccounts::<T>::insert(&address, &account_id);
+        EvmAccounts::<T>::insert(&account_id, &address);
+
+        Self::deposit_event(Event::AccountClaimed {
+            account_id,
+            evm_address: address,
+        });
+    }
+
+    /// Claim the given evm address by providing claim signature
+    fn do_claim_address(
+        account_id: T::AccountId,
+        evm_address: EvmAddress,
+        signature: SignatureOf<T>,
+    ) -> DispatchResult {
+        // recover evm address from signature
+        let address = T::ClaimSignature::verify_signature(&account_id, &signature)
+            .ok_or(Error::<T>::BadSignature)?;
+        log::trace!(
+            target: "account::do_claim_address",
+            "evm_address: {:#?}, recovered: {:#?}", evm_address, address
+        );
+        ensure!(evm_address == address, Error::<T>::InvalidSignature);
+
+        // Check if the default account id already exists for this eth address
+        let default_account_id = T::DefaultAddressMapping::into_account_id(evm_address.clone());
+        if frame_system::Pallet::<T>::account_exists(&default_account_id) {
+            // Transfer all the free balance from old account id to the newly
+            // since this `default_account_id` will no longer be connected to evm address
+            // and users cannot access it.
+            T::Currency::transfer(
+                &default_account_id,
+                &account_id,
+                T::Currency::reducible_balance(&default_account_id, Expendable, Polite),
+                Expendable,
+            )?;
+        }
+
+        // create double mappings for the pair
+        Self::add_mappings(account_id, evm_address);
+        Ok(())
+    }
+
+    /// Claim the default evm address
+    fn do_claim_default_address(account_id: T::AccountId) -> Result<EvmAddress, DispatchError> {
+        // get the default evm address
+        let address = T::DefaultAccountMapping::into_h160(account_id.clone());
+        // create double mappings for the pair with default evm address
+        Self::add_mappings(account_id, address.clone());
+        Ok(address)
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    #[cfg(any(feature = "std", feature = "runtime-benchmarks"))]
```

Comment:
I guess you could put this once, above the `impl`?

Also, by `"std"`, do you actually mean `test`?

Reply:
Nice catch, will do!
Yes it's for tests and benchmarks